middleware.go

```diff
@@ -131,3 +138,114 @@ func (lr roundTripRecorder) RoundTrip(req *http.Request) (*http.Response, error)
 
 	return res, err
 }
+
+// RetryMiddleware allows retrying of certain retryable http errors.
+// This only handles very specific status codes, ones that are deemed retryable;
+//
+//  - 502 Bad Gateway
+//  - 503 Service Unavailable
+//  - 504 Gateway Timeout
+type retryMiddleware struct {
+	policy              RetryPolicy
+	wrappedRoundTripper http.RoundTripper
+	clock               clock.Clock
+}
+
+type RetryPolicy struct {
+	Delay    time.Duration
+	Attempts int
+}
+
+// Validate validates the RetryPolicy for any issues.
+func (p RetryPolicy) Validate() error {
+	if p.Attempts < 1 {
+		return errors.Errorf("expected at least one attempt")
+	}
+	return nil
+}
+
+// makeRetryMiddleware creates a retry transport.
+func makeRetryMiddleware(transport http.RoundTripper, policy RetryPolicy, clock clock.Clock) http.RoundTripper {
+	return retryMiddleware{
+		policy:              policy,
+		wrappedRoundTripper: transport,
+		clock:               clock,
+	}
+}
+
+type retryableErr struct{}
+
+func (retryableErr) Error() string {
+	return "retryable error"
+}
+
+func (m retryMiddleware) RoundTrip(req *http.Request) (*http.Response, error) {
+	var res *http.Response
+	err := retry.Call(retry.CallArgs{
+		Clock: m.clock,
+		Func: func() error {
+			if err := req.Context().Err(); err != nil {
+				return err
+			}
+
+			var retryable bool
+			var err error
+			res, retryable, err = m.roundTrip(req)
+			if err != nil {
+				return err
+			}
+			if retryable {
+				return retryableErr{}
+			}
+			return nil
+		},
+		IsFatalError: func(err error) bool {
+			// Work out if it's not a retryable error.
+			_, ok := errors.Cause(err).(retryableErr)
+			return !ok
+		},
+		Attempts: m.policy.Attempts,
+		Delay:    m.policy.Delay,
+		BackoffFunc: func(delay time.Duration, attempts int) time.Duration {
+			return m.defaultBackoff(res, delay)
+		},
+	})
+
+	return res, err
+}
+
+func (m retryMiddleware) roundTrip(req *http.Request) (*http.Response, bool, error) {
+	res, err := m.wrappedRoundTripper.RoundTrip(req)
+	if err != nil {
+		return nil, false, err
+	}
+
+	switch res.StatusCode {
+	case http.StatusBadGateway, http.StatusGatewayTimeout:
+		// The request should be retryable.
+		fallthrough
+	case http.StatusServiceUnavailable, http.StatusTooManyRequests:
+		// The request should be retryable, but additionally should contain
+		// a potential Retry-After header.
+		return res, true, nil
+	default:
+		// Don't handle any of the following status codes.
+		return res, false, nil
+	}
+}
+
+// defaultBackoff attempts to workout a good backoff strategy based on the
+// backoff policy or the status code from the response. The
+func (m retryMiddleware) defaultBackoff(resp *http.Response, max time.Duration) time.Duration {
+	if resp.StatusCode == http.StatusServiceUnavailable || resp.StatusCode == http.StatusTooManyRequests {
```

Comment:
The RFC indicated that only 503 may return Retry-After, doesn't hurt to check others.